File: src/PY/srtf.py
def srtf(processos):
    lista = [
        {**p, "remaining": p["burst"], "color_idx": p.get("color_idx", 0)}
        for p in processos
    ]
    
    t = 0
    done = 0
    n = len(lista)
    gantt = []
    ct = {}
    
    current_pid = None
    block_start = 0
    current_color = 0
    
    while done < n:
        # Encontra todos os processos que já chegaram e ainda precisam executar
        available = [p for p in lista if p["arrival"] <= t and p["remaining"] > 0]
        
        if not available:
            if current_pid is not None:
                gantt.append({"pid": current_pid, "start": block_start, "end": t, "color_idx": current_color})
                current_pid = None
            
            # Pula o tempo para a próxima chegada
            t = min([p["arrival"] for p in lista if p["remaining"] > 0])
            continue
        
        # Pega o processo com menor tempo restante
        available.sort(key=lambda x: (x["remaining"], x["arrival"]))
        shortest = available[0]
        
        # Se houve troca de contexto, registra o bloco anterior no Gantt
        if shortest["pid"] != current_pid:
            if current_pid is not None:
                gantt.append({"pid": current_pid, "start": block_start, "end": t, "color_idx": current_color})
            current_pid = shortest["pid"]
            block_start = t
            current_color = shortest["color_idx"]
            
        # Executa por 1 unidade de tempo (preempção)
        shortest["remaining"] -= 1
        t += 1
        
        # Se finalizou
        if shortest["remaining"] == 0:
            gantt.append({"pid": current_pid, "start": block_start, "end": t, "color_idx": current_color})
            ct[shortest["pid"]] = t
            current_pid = None
            done += 1

    # Calcula as métricas exigidas
    metricas = []
    for p in processos:
        tat = ct[p["pid"]] - p["arrival"]
        wt  = tat - p["burst"]
        metricas.append({
            "pid": p["pid"], "arrival": p["arrival"], "burst": p["burst"],
            "ct": ct[p["pid"]], "tat": tat, "wt": wt,
            "color_idx": p.get("color_idx", 0),
        })

    tme = sum(m["wt"]  for m in metricas) / len(metricas) if metricas else 0
    tmt = sum(m["tat"] for m in metricas) / len(metricas) if metricas else 0
    
    return gantt, metricas, tme, tmt

File: src/JS/srtf.js
function srtf(procs) {
  const list = procs.map(p => ({...p, remaining: p.burst})).sort((a, b) => a.arrival - b.arrival);
  let t = 0, done = 0;
  const n = list.length;
  const gantt = [];
  const ct = {};

  let current_pid = null;
  let block_start = 0;
  let current_color = 0;

  while (done < n) {
      const available = list.filter(p => p.arrival <= t && p.remaining > 0);

      if (available.length === 0) {
          if (current_pid !== null) {
              gantt.push({pid: current_pid, start: block_start, end: t, colorIdx: current_color});
              current_pid = null;
          }
          const future = list.filter(p => p.remaining > 0);
          t = Math.min(...future.map(p => p.arrival));
          continue;
      }

      available.sort((a, b) => {
          if (a.remaining === b.remaining) return a.arrival - b.arrival;
          return a.remaining - b.remaining;
      });
      const shortest = available[0];

      if (shortest.pid !== current_pid) {
          if (current_pid !== null) {
              gantt.push({pid: current_pid, start: block_start, end: t, colorIdx: current_color});
          }
          current_pid = shortest.pid;
          block_start = t;
          current_color = shortest.colorIdx;
      }

      shortest.remaining -= 1;
      t += 1;

      if (shortest.remaining === 0) {
          gantt.push({pid: current_pid, start: block_start, end: t, colorIdx: current_color});
          ct[shortest.pid] = t;
          current_pid = null;
          done += 1;
      }
  }

  const metrics = procs.map(p => {
      const tat = ct[p.pid] - p.arrival;
      const wt = tat - p.burst;
      return {
          pid: p.pid, arrival: p.arrival, burst: p.burst,
          ct: ct[p.pid], tat: tat, wt: wt,
          colorIdx: p.colorIdx
      };
  });

  const tme = metrics.reduce((s, m) => s + m.wt, 0) / metrics.length || 0;
  const tmt = metrics.reduce((s, m) => s + m.tat, 0) / metrics.length || 0;

  return { gantt, metrics, tme, tmt, total: t };
}